scheduler/src/cook/hooks.clj

```diff
@@ -0,0 +1,174 @@
+          (do
+            (log/info (str "Resolved as " resolved-fn " with " arguments))
+            (resolved-fn arguments))
+          (throw (ex-info "Unable to resolve factory function" (assoc hook-factory :ns (namespace factory-fn))))))
+      accept-all-hook)))
+
+;  Contains the hook object that matches to a given job map. This code may create a new hook object or re-use an existing one.
+(mount/defstate hook-object
+  :start (create-default-hook-object config))
+
+(mount/defstate submission-hook-batch-timeout-seconds
+  :start (-> config :settings :hooks :submission-hook-batch-timeout-seconds t/seconds))
+(mount/defstate age-out-last-seen-deadline-minutes
+  :start (-> config :settings :hooks :age-out-last-seen-deadline-minutes t/minutes))
+(mount/defstate age-out-first-seen-deadline-minutes
+  :start (-> config :settings :hooks :age-out-first-seen-deadline-minutes t/minutes))
+(mount/defstate age-out-seen-count
+  :start (-> config :settings :hooks :age-out-seen-count))
+
+; We may see up to the entire scheduler queue, so have a big cache here.
+; This is called in the scheduler loop. If it hasn't been looked at in more than 2 hours, the job has almost assuredly long since run.
+(def ^Cache job-invocations-cache
+  (-> (CacheBuilder/newBuilder)
+      (.maximumSize 100000)
+      (.expireAfterAccess 2 TimeUnit/HOURS)
+      (.build)))
+
+
+(defn aged-out?
+  "Looking at the cached dictionary, Has this job been in the queue long enough that we should force it to run and get it done. This includes:
+  1. Being in the queue for a long time (hours)
+  2. Attempting to launch the job fairly recently (seen in last several minutess)
+  3. Tried multiple times to get the job to launch."
+  [{:keys [last-seen first-seen seen-count] :as old-result}]
+  {:post [(or (true? %) (false? %))]}
+  (let [last-seen-deadline (->> age-out-last-seen-deadline-minutes
+                                (t/minus- (t/now)))
+        first-seen-deadline (->> age-out-first-seen-deadline-minutes
+                                 (t/minus- (t/now)))]
+    (and
+      (boolean old-result)
+      (> seen-count age-out-seen-count)
+      (t/before? first-seen first-seen-deadline)
+      (t/before? last-seen-deadline last-seen))))
+
+
+(defn filter-job-invocations-miss
+  "This is the cache miss handler. It is invoked if we have a cache miss --- either the
+  entry is expired, or its not there. Only invoke on misses or expirations, because we
+  count the number of invocations."
+  [job]
+  {:post [(or (true? %) (false? %))]}
+  (let [{:keys [first-seen seen-count] :as old-result} (ccache/get-if-present
+                                                         job-invocations-cache
+                                                         :job/uuid
+                                                         job)
+        not-found? (not old-result)
+        is-aged-out? (aged-out? old-result)]
+    ; If aged-out, no more backend queries. Return true so the job is kept and
```

Comment:
The control flow comments are smelly to me. Can we rewrite this so it's more explicit in the actual structure of the code?